File: frontend/src/survey/completion.ts
import { ANSWERABLE, type Answers, type Definition } from "./types";
import { isAnswered, matrixRows, pendingKeys, questionByKey, visibleQuestions } from "./visibility";

/** Visible answerable questions without an answer row (RUN-18). */
export function missingKeys(def: Definition, answers: Answers): string[] {
  const questions = questionByKey(def);
  const missing: string[] = [];
  for (const v of visibleQuestions(def, answers)) {
    if (!ANSWERABLE.has(v.type)) continue;
    const value = answers[v.key];
    if (value === undefined) {
      missing.push(v.key);
      continue;
    }
    const q = questions[v.key];
    if (q.type === "matrix" && isAnswered(value) && "ratings" in value) {
      const rows = matrixRows(q, answers, questions);
      const rated = Object.keys(value.ratings);
      if (rows.length !== rated.length || !rows.every((r) => rated.includes(r))) missing.push(v.key);
    }
  }
  return missing;
}

/**
 * Progress over the whole instrument, so the total never moves. `total` counts
 * every answerable question in the definition, visible or not; `answered` is
 * everything settled — visible questions with an answer (a pruned matrix counts
 * as open, exactly as `missingKeys` reports it; a skip counts as answered) and
 * hidden questions whose branch is closed. Only hidden questions that may still
 * appear (`pendingKeys`) are left out. Mirrors completion.py.
 */
export function progress(def: Definition, answers: Answers): { answered: number; total: number } {
  const total = def.sections.flatMap((s) => s.questions).filter((q) => ANSWERABLE.has(q.type)).length;
  return { answered: total - missingKeys(def, answers).length - pendingKeys(def, answers).length, total };
}

File: backend/prolog_surveys/engine/completion.py
"""Completion rule (RUN-18): every visible answerable question has an answer row."""

from __future__ import annotations

from typing import Any

from .visibility import (
    ANSWERABLE,
    Answers,
    VisibleQuestion,
    is_answered,
    matrix_rows,
    question_by_key,
    visible_questions,
)


def missing_keys(
    definition: dict[str, Any],
    answers: Answers,
    *,
    visible: list[VisibleQuestion] | None = None,
    questions: dict[str, dict[str, Any]] | None = None,
) -> list[str]:
    """Visible answerable questions without a complete answer row.

    ``visible``/``questions`` accept a caller's precomputed walk so one request
    does not traverse the definition several times.
    """
    if questions is None:
        questions = question_by_key(definition)
    if visible is None:
        visible = visible_questions(definition, answers, questions=questions)
    missing: list[str] = []
    for v in visible:
        if v.type not in ANSWERABLE:
            continue
        value = answers.get(v.key)
        if value is None:
            missing.append(v.key)
            continue
        q = questions[v.key]
        if q["type"] == "matrix" and is_answered(value):
            rows = matrix_rows(q, answers, questions)
            if set(value.get("ratings", {})) != set(rows):
                missing.append(v.key)
    return missing


def progress(
    definition: dict[str, Any],
    answers: Answers,
    *,
    visible: list[VisibleQuestion] | None = None,
    missing: list[str] | None = None,
    questions: dict[str, dict[str, Any]] | None = None,
) -> dict[str, int]:
    """Answered = visible answerable questions that are not missing, so a
    pruned matrix (rated, but not for every current row) counts as open,
    exactly as ``missing_keys`` reports it; a skip counts as answered."""
    if visible is None:
        visible = visible_questions(definition, answers, questions=questions)
    if missing is None:
        missing = missing_keys(definition, answers, visible=visible, questions=questions)
    total = sum(1 for v in visible if v.type in ANSWERABLE)
    return {"answered": total - len(missing), "total": total}
